Refactor dynamo distributed test helpers to be reusable

wconstab opened the pull request:
The point is to let Test helpers previously defined and used in `test_dynamo_distributed.py` be used from a new file `test_traceable_collectives.py` later in this stack.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #93113
* #93112
* #93111
* #93110
* #93109
* __->__ #93187

wconstab committed:
Refactor dynamo distributed test helpers to be reusable

[ghstack-poisoned]

kumpera reviewed (approved):
LGTM.

But please do not use MultiProcessTestCase unless you must. It's super super slow and we're trying to migrate the test suite to our multi threaded PG:

See https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/distributed/multi_threaded_pg.py

wconstab committed:
Update on "Refactor dynamo distributed test helpers to be reusable"

The point is to let Test helpers previously defined and used in `test_dynamo_distributed.py` be used from a new file `test_traceable_collectives.py` later in this stack.




[ghstack-poisoned]

wconstab committed:
Update on "Refactor dynamo distributed test helpers to be reusable"

The point is to let Test helpers previously defined and used in `test_dynamo_distributed.py` be used from a new file `test_traceable_collectives.py` later in this stack.




[ghstack-poisoned]

wconstab committed:
Update on "Refactor dynamo distributed test helpers to be reusable"

The point is to let Test helpers previously defined and used in `test_dynamo_distributed.py` be used from a new file `test_traceable_collectives.py` later in this stack.




[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge -f "flaky ci (windows)"
